### PR TITLE
build: add a CMake based build

### PR DESCRIPTION
**Pull request by compnerd**

### CODE REVIEW

**Comment by compnerd** [comment]
CC: @pschuh

File: README.md [hunk 1]
```diff
@@ -84,13 +84,27 @@ For more models, go to [**tensorflow/swift-models**](https://github.com/tensorfl
...
 ```
 $ swift build
 ```
 ```
 $ swift test
 ```
 
+#### Building using CMake
+
+Note that this requires CMake 3.15.1 or newer.  In-tree builds are not
```
Comment (dan-zheng) [inline README.md]: I wonder what "in-tree build" means? I found https://unix.stackexchange.com/a/208643 but it doesn't quite explain.

File: README.md [hunk 2]
```diff
@@ -84,13 +84,27 @@ For more models, go to [**tensorflow/swift-models**](https://github.com/tensorfl
...
+supported.  The following assumes that the intermediates should be placed in
+`/BinaryCache/tensorflow-swift-apis` and that the sources have been checked out
```
Comment (dan-zheng) [inline README.md]: To build using CMake, is it required for `tensorflow-swift-apis` sources to exist at `/SourceCache/tensorflow-swift-apis`?

Can sources and binary intermediates be placed elsewhere? It's not clear from the current instructions.

Comment (compnerd) [reply README.md]: They can be placed anywhere.  I can reword the "assumes that the intermediates should placed in ... and that the sources have been checked out at ..."

File: README.md (hunk 1 above)
Comment (compnerd) [reply README.md]: It just means that the build intermediates and artifacts are in the source tree.

**Comment by dan-zheng** [comment]
I encountered this error:
```console
$ cmake -B out -S . -D CMAKE_Swift_COMPILER=/Users/danielzheng/.swiftenv/shims/swiftc
CMake Error at /usr/local/Cellar/cmake/3.15.5/share/cmake/Modules/CMakeDetermineSwiftCompiler.cmake:45 (message):
  Swift language not supported by "Unix Makefiles" generator
Call Stack (most recent call first):
  CMakeLists.txt:2 (project)


-- Configuring incomplete, errors occurred!
See also "/Users/danielzheng/swift-apis/out/CMakeFiles/CMakeOutput.log".
```

Do you have any idea what's the cause?

**Comment by compnerd** [comment]
Ah, add `-G Ninja`

**Comment by dan-zheng** [comment]
> Ah, add `-G Ninja`

Thanks! That seems to work:
```console
$ cmake -G Ninja -B out -S . -D CMAKE_Swift_COMPILER=/Users/danielzheng/.swiftenv/shims/swiftc
-- The CXX compiler identification is AppleClang 11.0.0.11000033
-- The Swift compiler identification is Apple 5.2
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working Swift compiler: /Users/danielzheng/.swiftenv/shims/swiftc
-- Check for working Swift compiler: /Users/danielzheng/.swiftenv/shims/swiftc -- works
CMake Error at /usr/local/Cellar/cmake/3.15.5/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find TensorFlow (missing: TensorFlow_INCLUDE_DIR
  TensorFlow_LIBRARY)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.15.5/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/FindTensorFlow.cmake:13 (find_package_handle_standard_args)
  CMakeLists.txt:9 (find_package)


-- Configuring incomplete, errors occurred!
See also "/Users/danielzheng/swift-apis/out/CMakeFiles/CMakeOutput.log".
See also "/Users/danielzheng/swift-apis/out/CMakeFiles/CMakeError.log".
```

This "TensorFlow not found" error seems expected. I wonder how to installed TensorFlow so that it can be found by CMake?

**Comment by compnerd** [comment]
You can install it into `/usr/lib` and the headers into `/usr/include` (or `/usr/local/lib` and `/usr/local/include`).  Alternatively, you can specify the paths manually using `-D TensorFlow_LIBRARY=/Library/tensorflow-2.1.0-rc1/usr/lib/libtensorflow.dylib` and `-D TensorFlow_INCLUDE_DIR=/Library/tensorflow-2.1.0-rc1/usr/include`.

**Comment by dan-zheng** [comment]
> -D TensorFlow_INCLUDE_DIR=

Thanks! I ran:

```console
```

**Comment by compnerd** [comment]
Yes, you need to do `cmake --build out` for it to do the build

**Comment by dan-zheng** [comment]
> Yes, you need to do `cmake --build out` for it to do the build

I see. I ran `cmake --build .` in the `out` directory:
```console
$ cmake --build .
[1/2] Linking Swift shared library Sources/Tensor/libTensor.dylib
FAILED: Sources/Tensor/libTensor.dylib Sources/Tensor/CMakeFiles/Tensor.dir/Random.swift.o Sources/Tensor/CMakeFiles/Tensor.dir/TensorUtilities.swift.o swift/Tensor.swiftmodule
: && /Users/danielzheng/.swiftenv/shims/swiftc -output-file-map Sources/Tensor/CMakeFiles/Tensor.dir/output-file-map.json -incremental -num-threads 16 -emit-library -o Sources/Tensor/libTensor.dylib -module-name Tensor -module-link-name Tensor -emit-module -emit-module-path swift/Tensor.swiftmodule -emit-dependencies -DTensor_EXPORTS   ../Sources/Tensor/Random.swift ../Sources/Tensor/TensorUtilities.swift      && :
<unknown>:0: error: unable to load output file map 'Sources/Tensor/CMakeFiles/Tensor.dir/output-file-map.json': No such file or directory
ninja: build stopped: subcommand failed.
```

Is this currently expected?

**Comment by compnerd** [comment]
No, thats not expected.  I think that you may want to update your CMake.  I've been testing with 3.16.

**Comment by dan-zheng** [comment]
> No, thats not expected. I think that you may want to update your CMake. I've been testing with 3.16.

I upgraded CMake `3.15.5 ->3.16.2` via `brew upgrade cmake`. That fixes the error above - shall we bump the minimum CMake version in the README?

Using an untouched [Swift for TensorFlow macOS toolchain](https://storage.googleapis.com/swift-tensorflow/mac/swift-tensorflow-DEVELOPMENT-2019-12-23-a-osx.pkg), I get a duplicate CTensorFlow module error:
```console
$ cmake --build . # out directory
...
/Users/danielzheng/swift-apis/out/../Sources/CTensorFlow/module.modulemap:1:8: error: redefinition of module 'CTensorFlow'
module CTensorFlow {
       ^
/Library/Developer/Toolchains/swift-tensorflow-DEVELOPMENT-2019-12-23-a.xctoolchain/usr/lib/swift/macosx/x86_64/modulemaps/CTensorFlow/module.modulemap:1:8: note: previously defined here
module CTensorFlow {
       ^
ninja: build stopped: subcommand failed.
```

---

Using a copied toolchain where I deleted the `CTensorFlow` modulemap in the toolchain, I get the following error:
```console
$ cmake --build . # out directory
...
<module-includes>:1:9: note: in file included from <module-includes>:1:
#import "CTensorFlow.h"
        ^
/Users/danielzheng/swift-apis/out/../Sources/CTensorFlow/CTensorFlow.h:4:10: error: 'tensorflow/c/c_api.h' file not found
#include <tensorflow/c/c_api.h>
         ^
../Sources/TensorFlow/Bindings/EagerExecution.swift:17:8: error: could not build Objective-C module 'CTensorFlow'
import CTensorFlow
       ^
ninja: build stopped: subcommand failed.
```

Any clue why the `tensorflow/c/c_api.h` header can't be found? Does building with CMake fully work for you?

**Comment by compnerd** [comment]
Yes, it does work for me (although, I've been building on Windows).  I'll do the update to 3.16 in a subsequent change.

As to why the `tensorflow/c/c_api.h` is not found ... your `-D TensorFlow_INCLUDE_DIR=~/swift-build/tensorflow/tensorflow/c` is the culprit ... you are two levels too deep.

Interesting point about the integrated copy.  I think that is something that is beyond the scope of this change and something we need to figure out to enable people iterate on this library without a custom build of TensorFlow.

**Pull request merged**